File: Design-Patterns-Using-Cpp/SOLID/LiskovSubstitutionPrinciple3.cpp
#include <iostream>

class Rectangle {
private:
    int height;
    int width;

public:
    Rectangle(int l, int w) : height(l), width(w) {}

    virtual void setWidth(int w) {
        width = w;
    }

    virtual void setHeight(int h) {
        height = h;
    }

    int getWidth() const {
        return width;
    }

    int getHeight() const {
        return height;
    }
};

class Square : public Rectangle {
public:
    Square(int l) : Rectangle(l, l) {}

    void setWidth(int w) override {
        Rectangle::setWidth(w);
        Rectangle::setHeight(w);
    }

    void setHeight(int h) override {
        Rectangle::setWidth(h);
        Rectangle::setHeight(h);
    }
};

void testRect(Rectangle& rect) {
    rect.setHeight(10);
    rect.setWidth(20);
    if (200 == rect.getHeight() * rect.getWidth())
        std::cout << "success" << std::endl;
    else
        std::cout << "failure" << std::endl;
}

// Client code
int main() {
    Rectangle r(10, 20);
    testRect(r);

    Square s(10);
    s.setWidth(20);
    testRect(s);

    return 0;
}

/*
success
failure
*/

File: Design-Patterns-Using-Ruby/SOLID/LiskovSubstitutionPrinciple3.rb
class Rectangle
    attr_accessor :height, :width

    def initialize(l, w)
        @height = l
        @width = w
    end

    def set_width(w)
        @width = w
    end
    
    def set_height(h)
        @height = h
    end

    def get_width
        @width
    end
    
    def get_height
        @height
    end
end

class Square < Rectangle
    def initialize(l)
        super(l, l)
    end

    def set_width(w)
        @width = w
        @height = w
    end

    def set_height(h)
        @height = h
        @width = h
    end
end

# Client code
def test_rect(rect)
    rect.set_height(10)
    rect.set_width(20)
    if 200 == rect.get_height * rect.get_width
        puts 'success'
    else
        puts 'failure'
    end
end


r = Rectangle.new(10, 20)
test_rect(r)

s = Square.new(10)
s.set_width(20)
test_rect(s)
